salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,535 @@
+        return '/'.join([self.group, self.version])
+
+    @property
+    def key(self):
+        # Used for indexing in `KNOWN_CUSTOM_OBJECTS`
+        return (self.api_version, self.kind)
+
+    def _method(self, verb):
+        return '{}_{}_custom_object'.format(verb, self.scope.value)
+
+
+_CUSTOM_OBJECTS = [
+    CRDetails('monitoring.coreos.com/v1', 'Alertmanager',
+              scope='namespaced', plural='alertmanagers'),
+    CRDetails('monitoring.coreos.com/v1', 'Prometheus',
+              scope='namespaced', plural='prometheuses'),
+    CRDetails('monitoring.coreos.com/v1', 'PrometheusRule',
+              scope='namespaced', plural='prometheusrules'),
+    CRDetails('monitoring.coreos.com/v1', 'ServiceMonitor',
+              scope='namespaced', plural='servicemonitors'),
+]
+
+KNOWN_CUSTOM_OBJECTS = {details.key: details for details in _CUSTOM_OBJECTS}
+
+
+class _DictWrapper(object):
+    """Wrapper for dynamic attribute access support over a simple dict.
+
+    Used by `CustomObject` to emulate the other models for `kubernetes.client`.
+    """
+    def __init__(self, fields):
+        self._fields = {
+            key: self.from_value(value) for key, value in fields.items()
+        }
+
+    @classmethod
+    def from_value(cls, value):
+        if isinstance(value, dict):
+            return cls(value)
+        if isinstance(value, list):
+            return [cls.from_value(val) for val in value]
+        return value
+
+    def __repr__(self):
+        return repr(self._fields)
+
+    def __getattribute__(self, name):
+        try:
+            return super(_DictWrapper, self).__getattribute__(name)
+        except AttributeError:
+            # Fallback on `_fields`
+            pass
+
+        try:
+            return self._fields[name]
+        except KeyError:
+            raise AttributeError(
+                "Custom object has no attribute '{}'".format(name)
+            )
+
+
+class CustomObject(object):
+    """Helper class to generate a "model" interface for CustomResources."""
+    def __init__(self, cr_details, manifest):
+        self._details = cr_details
+        self._body = _cast_dict_keys(manifest, _camel_to_snake)
+        self._attr_dict = _DictWrapper(self._body)
+
+    details = property(operator.attrgetter('_details'))
+    body = property(operator.attrgetter('_body'))
+
+    def to_dict(self):
+        return self.body
+
+    def to_str(self):
+        return pformat(self.to_dict())
+
+    def __repr__(self):
+        return self.to_str()
+
+    def __eq__(self, other):
+        if not isinstance(other, CustomObject):
+            return False
+
+        return self.body == other.body
+
+    def __getattribute__(self, name):
+        try:
+            return super(CustomObject, self).__getattribute__(name)
+        except AttributeError:
+            return getattr(self._attr_dict, name)
+
+
+def __virtual__():
+    return __virtualname__
+
+
+def get_object_details(manifest):
```

Comment:
Nit: the naming is a bit confusing/conflates two concepts: it *takes* a `manifest`, which is an object (think 'instance'), but it doesn't return any kind of 'details' about this object, instead, it returns *meta-information* about the *type* of the object.

Reply:
Agreed, I'm thinking these `ObjectDetails` / `CRDetails` could be renamed to something like `Kind` / `CustomKind`, though I'm not super fond of it either... Any suggestions?